src/procedures/proc_bfs.c

```diff
@@ -0,0 +1,222 @@
+	GrB_Vector parents;             // Vector associating each node in the BFS tree with its parent.
+	int reltype_id;                 // ID of relationship matrix to traverse.
+	int nodes_output_idx;           // Offset of nodes array in outputs
+	int edges_output_idx;           // Offset of edges array in outputs
+	SIValue *output;                // Array with a maximum of 4 entries: ["nodes", nodes, "edges", edges].
+	bool depleted;                  // True if BFS has already been performed for this node.
+} BFSContext;
+
+static ProcedureResult Proc_BFS_Invoke(ProcedureCtx *ctx, const SIValue *args) {
+	if(array_len((SIValue *)args) != 4) return PROCEDURE_ERR;
+	if(SI_TYPE(args[0]) != T_NODE                 ||   // Source node.
+	   SI_TYPE(args[1]) != T_INT64                ||   // Max level to iterate to, unlimited if 0.
+	   !(SI_TYPE(args[2]) & (T_NULL | T_STRING))  ||   // Relationship type to traverse if not NULL.
+	   SI_TYPE(args[3]) != T_BOOL)                     // Whether traversed edges should be returned.
+		return PROCEDURE_ERR;
+
+	BFSContext *pdata = ctx->privateData;
+	pdata->depleted = false;
+	Node *source_node = args[0].ptrval;
+	GrB_Index source_id = ENTITY_GET_ID(source_node);
+	int64_t max_level = args[1].longval;
+	/* The BFS algorithm uses a level of 1 to indicate the source node. If this value is not
+	 * zero (unlimited), increment it by 1 to make level 1 indicate the source's direct neighbors. */
+	if(max_level > 0) max_level++;
+	const char *reltype = SIValue_IsNull(args[2]) ? NULL : args[2].stringval;
+	bool collect_edges = args[3].longval;
+
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	// Get edge matrix and transpose matrix, if available.
+	GrB_Matrix R;
+	GrB_Matrix TR;
+	if(reltype == NULL) {
+		R = Graph_GetAdjacencyMatrix(gc->g);
+		TR = Graph_GetTransposedAdjacencyMatrix(gc->g);
+	} else {
+		Schema *s = GraphContext_GetSchema(gc, reltype, SCHEMA_EDGE);
+		if(!s) return PROCEDURE_OK; // Failed to find schema, first step will return NULL.
+		pdata->reltype_id = s->id;
+		R = Graph_GetRelationMatrix(gc->g, s->id);
+		if(Config_MaintainTranspose()) TR = Graph_GetTransposedRelationMatrix(gc->g, s->id);
+		else TR = GrB_NULL;
+	}
+
+	GrB_Vector V;  // Vector of results
+	GrB_Vector PI = NULL; // Vector backtracking results to their parents.
+	GrB_Vector *PI_ptr = &PI;
+	if(!collect_edges) {
+		/* If we're not collecting edges, pass a NULL parent pointer so that the algorithm will
+		 * not perform unnecessary work. */
+		PI_ptr = GrB_NULL;
+		pdata->edges_output_idx = -1;
+	}
+	assert(LAGraph_bfs_both(&V, PI_ptr, R, TR, source_id, max_level, false) == GrB_SUCCESS);
+	/* Remove all values with a level less than or equal to 1.
+	 * Values of 0 are not connected to the source, and values of 1 are the source. */
+	GxB_Scalar thunk;
+	GxB_Scalar_new(&thunk, GrB_UINT64);
+	GxB_Scalar_setElement_UINT64(thunk, 1);
+	GxB_Vector_select(V, GrB_NULL, GrB_NULL, GxB_GT_THUNK, V, thunk, GrB_NULL);
+	GxB_Scalar_free(&thunk);
```

Comment:
If I understand correctly, `false` means we expect `V` to be dense:
```
bool vsparse            // if true, v is expected to be very sparse
```
This seems like the right call to me for this algorithm, but let me know if you think it should be changed.

Reply:
If i understand correctly,because u expect v is dense,so in algorithm v has been converted to dense.It is not necessary to remove values less than or equal to 1.

Reply:
For some additional context on vsparse:
```
//      bool vsparse:  if the result v may remain very sparse, then set this
//          parameter to true.  If v might have many entries, set it false.  If
//          you are unsure, then set it to true.  This parameter speeds up
//          the handling of v.  If you guess wrong, there is a slight
//          performance penalty.  The results are not affected by this
//          parameter, just the performance.  This parameter is used only for
//          the single-source BFS.
```
I'm unclear on when the vector could _ever_ remain sparse, because when examining the tuples I found explicit 0 values for every unconnected element, which is to say that the vector was fully populated. It's for this reason that I'm removing values <= 1, so that we can iterate over elements that solely represent descendants on the BFS tree.